crates/chia-datalayer/src/lib.rs

```diff
@@ -1,42 +1,155 @@
-use std::collections::HashMap;
+// use std::collections::HashMap;
 
-type Index = usize;
-type Key = Vec<u8>;
+type TreeIndex = u32;
+// type Key = Vec<u8>;
+type Hash = [u8; 32];
+type KVId = Hash;
+
+#[derive(Debug, Hash, Eq, PartialEq)]
 
 pub enum NodeType {
     Internal,
     Leaf,
 }
 
+impl NodeType {
+    pub fn load(&value: &u8) -> Self {
+        // TODO: identify some useful structured serialization tooling we use
+        // TODO: find a better way to tie serialization values to enumerators
+        match value {
+            0 => NodeType::Internal,
+            1 => NodeType::Leaf,
+            other => panic!("unknown NodeType value: {}", other),
+        }
+    }
+}
+
+// impl NodeType {
+//     const TYPE_TO_VALUE: HashMap<NodeType, u8> = HashMap::from([
+//         (NodeType::Internal, 0),
+//         (NodeType::Leaf, 1),
+//     ]);
+//
+//     fn value(&self) -> u8 {
+//         let map = Self::TYPE_TO_VALUE;
+//         // TODO: this seems pretty clearly the wrong way, probably
+//         let value = map.get(self);
+//         if value.is_some() {
+//             return 3;
+//         }
+//         panic!("no value for NodeType: {self:?}");
+//     }
+// }
+
 #[derive(Debug)]
 pub struct MerkleBlob {
     // TODO: shouldn't really all be pub
     pub blob: Vec<u8>,
-    pub kv_to_index: HashMap<Key, Index>,
-    pub free_indexes: Vec<Index>,
-    pub last_allocated_index: Index,
 }
 
+// TODO: fill out related to the serializations
+const METADATA_SIZE: u32 = 2;
+const DATA_SIZE: u32 = 0;
+const SPACING: u32 = METADATA_SIZE + DATA_SIZE;
+
+impl MerkleBlob {
+    pub fn get_raw_node(&self, index: TreeIndex) -> RawMerkleNode {
+        // TODO: handle invalid indexes?
+        // TODO: handle overflows?
```

Comment:
Overflows and invalid indices will cause a panic which will raise an error in Python and crash in Rust by default. You can instead use `.get(index)` to get an `Option` for example, and handle errors that way.

Reply:
I believe overflows only panic in debug builds?

Is indexing a case where I'm allowed to code a thing that might panic and I'm not given any warning to handle it instead?  So I should just always `.get()` instead of `[]` for indexing?  Though in this case the 'indexes' are used for slicing so I guess I should check how slices are handled and how to get them more safely.

Reply:
Overflows definitely always panic, otherwise it would be undefined behavior and a memory safety issue.
There are unchecked versions that don't panic, but they are marked as `unsafe` for those reasons.

I'd say if you don't know the length beforehand, just don't use indexing yeah.